File: backend/app/styles/catalog.py
"""Style catalog — loaded from recipes.json (static reference data).

Each recipe is a hand-/AI-authored prompt recipe. Internal prompt cues
(positive_cues / negative_cues / base_qualifiers) drive generation and are NOT
exposed to the client (the "secret sauce"); only public fields are served.
"""

import json
from dataclasses import dataclass
from functools import lru_cache
from pathlib import Path

_RECIPES_PATH = Path(__file__).parent / "recipes.json"


@dataclass(frozen=True)
class StyleRecipe:
    slug: str
    name: str
    category: str
    description: str
    positive_cues: str
    negative_cues: str
    base_qualifiers: str
    default_modifiers: dict
    tags: list[str]


@lru_cache
def load_recipes() -> dict[str, StyleRecipe]:
    data = json.loads(_RECIPES_PATH.read_text(encoding="utf-8"))
    rows = data["recipes"] if isinstance(data, dict) else data
    out: dict[str, StyleRecipe] = {}
    for r in rows:
        out[r["slug"]] = StyleRecipe(
            slug=r["slug"],
            name=r["name"],
            category=r.get("category", "other"),
            description=r.get("description", ""),
            positive_cues=r.get("positive_cues", ""),
            negative_cues=r.get("negative_cues", ""),
            base_qualifiers=r.get("base_qualifiers", ""),
            default_modifiers=r.get("default_modifiers", {}),
            tags=r.get("tags", []),
        )
    return out


def all_recipes() -> list[StyleRecipe]:
    return sorted(load_recipes().values(), key=lambda r: (r.category, r.name))


def get_recipe(slug: str) -> StyleRecipe | None:
    return load_recipes().get(slug)

File: worker/src/styles/catalog.ts
/**
 * Style catalog — loaded from recipes.json (static reference data).
 *
 * Internal prompt cues (positive_cues / negative_cues / base_qualifiers) drive
 * generation and are NOT exposed to the client; only public fields are served.
 */

import data from "./recipes.json";

export interface StyleRecipe {
  slug: string;
  name: string;
  category: string;
  description: string;
  positive_cues: string;
  negative_cues: string;
  base_qualifiers: string;
  default_modifiers: Record<string, unknown>;
  tags: string[];
}

type RawRecipe = Partial<StyleRecipe> & { slug: string; name: string };

const RECIPES = new Map<string, StyleRecipe>();
for (const r of ((data as { recipes: RawRecipe[] }).recipes ?? []) as RawRecipe[]) {
  RECIPES.set(r.slug, {
    slug: r.slug,
    name: r.name,
    category: r.category ?? "other",
    description: r.description ?? "",
    positive_cues: r.positive_cues ?? "",
    negative_cues: r.negative_cues ?? "",
    base_qualifiers: r.base_qualifiers ?? "",
    default_modifiers: r.default_modifiers ?? {},
    tags: r.tags ?? [],
  });
}

// Python sorts by (category, name) with plain code-point comparison.
const cmp = (a: string, b: string) => (a < b ? -1 : a > b ? 1 : 0);
const SORTED = [...RECIPES.values()].sort((a, b) => cmp(a.category, b.category) || cmp(a.name, b.name));

export function allRecipes(): StyleRecipe[] {
  return SORTED;
}

export function getRecipe(slug: string): StyleRecipe | undefined {
  return RECIPES.get(slug);
}
